### PR TITLE
feat: support Qwen/Qwen3-Omni-30B-A3B-Instruct and Audiomcq dataset.

### PR DESCRIPTION
**Pull request by yuekaizhang**
# What does this PR do ?
Conditioned on https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3317

The PR adds support for the ICLR 2026 [AudioMCQ](https://huggingface.co/datasets/inclusionAI/AudioMCQ) datasets in nemo-rl and includes a qwen2.5-omni-7B and Qwen3-Omni audio grpo training YAML file using these datasets.

Training curve:

image

@yuki-97  I was wondering if you coule help review the PR when you have some time. Thanks.

### CODE REVIEW

**Comment by yuekaizhang** [comment]
Update: added Qwen/Qwen3-Omni-30B-A3B-Instruct audio grpo

AudioMCQ training curve:
image

File: docs/guides/grpo-audio.md [hunk 1]
```diff
@@ -1,81 +1,196 @@
...
+| `cluster.num_nodes` × `gpus_per_node` | 4 × 8 |
+| Megatron `tensor_model_parallel_size` / `expert_model_parallel_size` / `pipeline_model_parallel_size` | 1 / 8 / 1 |
+| Megatron `moe_token_dispatcher_type` | allgather (matches EP≥16 large-MoE peers) |
+| vLLM `tensor_parallel_size` / `expert_parallel_size` / `async_engine` | 4 / 8 / false |
+| `train_global_batch_size` | 32 |
+| `max_total_sequence_length` | 2048 |
+| Learning rate | 1e-6 |
+| Reward | format (0.2) + exact_alnum (0.8) |
+
+The Qwen3-Omni recipe has three model-specific gotchas baked into the yaml:
```
Comment (yuki-97) [inline docs/guides/grpo-audio.md]:
```suggestion
The Qwen3-Omni recipe has two model-specific gotchas baked into the yaml:
```

File: examples/configs/recipes/vlm/vlm_grpo-qwen3-omni-30ba3b-audiomcq-4n8g-megatron.v1.yaml
Comment (yuki-97) [inline examples/configs/recipes/vlm/vlm_grpo-qwen3-omni-30ba3b-audiomcq-4n8g-megatron.v1.yaml:1]: let's remove this file and move other two yamls to `examples/configs/recipes/vlm/` (+ adding them to nightly), you can set max_num_steps to 20 in the nightly script.
docs may need related changes as well.

File: examples/configs/audio_grpo_qwen3_omni_megatron.yaml [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+      # topology to actually run with experts split across ranks. NeMo-RL
+      # reads vllm_cfg.expert_parallel_size at runtime (not megatron_cfg).
+      # On 4x8 (32 GPUs), EP=32 / TP=4 gives DP=8 vLLM groups each TP=4.
+      expert_parallel_size: 8
+      # async_engine must be false when ep_size != tp_size (vLLM DP path).
+      async_engine: false
+      gpu_memory_utilization: 0.55
+      max_model_len: ${policy.max_total_sequence_length}
+      limit_mm_per_prompt:
+        audio: 1
```
Comment (yuki-97) [inline examples/configs/audio_grpo_qwen3_omni_megatron.yaml]: here seems useless? looks already defined under `vllm_kwargs`.
```suggestion
```

File: examples/configs/audio_grpo_7B_megatron.yaml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+  generation:
+    max_new_tokens: 1024
+    vllm_cfg:
+      # Audio/multimodal models require tokenizer to be initialized before generation
+      skip_tokenizer_init: False
+      tensor_parallel_size: 2
+      gpu_memory_utilization: 0.55
+      limit_mm_per_prompt:
+        audio: 1
```
Comment (yuki-97) [inline examples/configs/audio_grpo_7B_megatron.yaml]: is this useful in this yaml? if so, need to move to under `vllm_kwargs` since put it here under `vllm_cfg` have no effects. if not, we can just remove it.

File: nemo_rl/data/datasets/response_datasets/audiomcq.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+                # so the two splits remain disjoint, but leave val_dataset
+                # unset so setup_response_data does not duplicate the rows
+                # the validation entry already contributes.
+                if split_validation_size > 0:
+                    split_ds = ds.train_test_split(
+                        test_size=split_validation_size, seed=seed
+                    )
+                    self.dataset = split_ds["train"]
+                else:
+                    self.dataset = ds
```
Comment (yuki-97) [inline nemo_rl/data/datasets/response_datasets/audiomcq.py]: curious why we need special handler here? is it possible to reuse the flow like other datasets like `nemo_rl/data/datasets/response_datasets/avqa.py`?

File: examples/configs/audio_grpo_qwen3_omni_megatron.yaml [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+data:
+  num_workers: 0
+  train:
+    dataset_name: audiomcq
+    split: train
+    # Held-out validation slice is synthesized from this train manifest.
+    # populate_val_dataset=false suppresses train-side auto-population so
+    # setup_response_data only sees the explicit validation entry below.
+    split_validation_size: 0.026
```
Comment (yuki-97) [inline examples/configs/audio_grpo_qwen3_omni_megatron.yaml]: related to the comments in dataset, also curious why we need set such small number of `split_validation_size` for all the recipes?

**Review by yuki-97 — COMMENTED** [review]
@yuekaizhang, thanks for adding this! left some comments.

File: docs/guides/grpo-audio.md (hunk 1 above)
Comment (yuekaizhang) [reply docs/guides/grpo-audio.md]: Fixed — there were only two bullets, updated the count to two.

File: examples/configs/audio_grpo_qwen3_omni_megatron.yaml (hunk 2 above)
Comment (yuekaizhang) [reply examples/configs/audio_grpo_qwen3_omni_megatron.yaml]: Removed. vllm.LLM() is built only from explicit fields + **vllm_kwargs, so the vllm_cfg copy was dead; kept the one under vllm_kwargs.

File: examples/configs/audio_grpo_7B_megatron.yaml (hunk 3 above)
Comment (yuekaizhang) [reply examples/configs/audio_grpo_7B_megatron.yaml]: Moved to vllm_kwargs — the audio cap is meaningful and only **vllm_kwargs reaches vllm.LLM().

File: nemo_rl/data/datasets/response_datasets/audiomcq.py (hunk 4 above)
Comment (yuekaizhang) [reply nemo_rl/data/datasets/response_datasets/audiomcq.py]: Simplified to mirror avqa.py: dropped populate_val_dataset and the split="validation" branch. The held-out slice is now synthesized via split_train_validation and exposed through val_dataset, so no separate data.validation entry is needed.

File: examples/configs/audio_grpo_qwen3_omni_megatron.yaml (hunk 5 above)
Comment (yuekaizhang) [reply examples/configs/audio_grpo_qwen3_omni_megatron.yaml]: Resolved with the dataset simplification — no more fractional split. Both recipes hold out an absolute 256 rows (= grpo.max_val_samples) via split_validation_size; the rest is training data.

File: examples/configs/recipes/vlm/vlm_grpo-qwen3-omni-30ba3b-audiomcq-4n8g-megatron.v1.yaml
Comment (yuekaizhang) [reply examples/configs/recipes/vlm/vlm_grpo-qwen3-omni-30ba3b-audiomcq-4n8g-megatron.v1.yaml:1]: Done: deleted the thin wrapper; moved both configs into examples/configs/recipes/vlm/ (vlm_grpo-qwen2.5-omni-7b-audiomcq-1n8g-megatron.v1.yaml and vlm_grpo-qwen3-omni-30ba3b-audiomcq-4n8g-megatron.v1.yaml, defaults -> ../../grpo_math_1B_megatron.yaml, minimized), added a 7B 1n8g nightly + registered both in nightly.txt (max_num_steps=20 set in the .sh), and updated the docs. Heads-up: after the main merge the branch exceeds the 1800 GPU-hour cap (1804 even without the 7B nightly, from the new TQ tests) — will resolve separately.

**Comment by yuekaizhang** [comment]
> @yuekaizhang, thanks for adding this! left some comments.

Many thanks @yuki-97!  All issues have been addressed.

File: nemo_rl/data/datasets/response_datasets/audiomcq.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        # both training and the held-out validation slice.
+        self.val_dataset = None
+        # split_validation_size is either a fraction in (0, 1) or an absolute
+        # row count. The typed config delivers an absolute count as a float
+        # (e.g. 256.0); datasets.train_test_split only accepts an int for
+        # absolute counts, so coerce whole numbers >= 1 back to int.
+        val_size = split_validation_size
+        if val_size >= 1:
+            val_size = int(val_size)
+        self.split_train_validation(val_size, seed)
```
Comment (yuki-97) [inline nemo_rl/data/datasets/response_datasets/audiomcq.py]: nit: maybe no need to convert after changing `split_validation_size` to `float | int`?
```suggestion
        self.split_train_validation(val_size, seed)
```

File: nemo_rl/data/datasets/response_datasets/audiomcq.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    pointing to a ``.wav`` or ``.mp3`` file shipped inline in the dataset
+    snapshot, plus a four-item ``choices`` list and a free-text ``answer``.
+    """
+
+    task_name = "audiomcq"
+
+    def __init__(
+        self,
+        split: str = "train",
+        split_validation_size: float = 0,
```
Comment (yuki-97) [inline nemo_rl/data/datasets/response_datasets/audiomcq.py]: oh, just find `datasets.Dataset.train_test_split` could support both 0~1 for fraction and >1 for absolute value.

nit:
```suggestion
        split_validation_size: float | int = 0,
```

also could you help also update `def split_train_validation(self, test_size: float, seed: int):` in `nemo_rl/data/datasets/raw_dataset.py` to `test_size: float | int`? thanks!

Comment (yuekaizhang) [reply nemo_rl/data/datasets/response_datasets/audiomcq.py]: Done in 724d82b. Widened to `float | int` here and updated `split_train_validation(test_size: float | int, seed)` in raw_dataset.py.

Kept the whole-number -> int coercion but moved it into `split_train_validation`: the recipes inherit a float default (`grpo_math_1B.yaml` -> `split_validation_size: 0.05`), so OmegaConf merges the int recipe value `256` into the float node and delivers `256.0` at runtime (confirmed in run.log). `train_test_split` rejects a float >= 1 as an out-of-range fraction. Centralizing the coercion means absolute counts now work for every dataset, not just AudioMCQ.

File: nemo_rl/data/datasets/response_datasets/audiomcq.py (hunk 6 above)
Comment (yuekaizhang) [reply nemo_rl/data/datasets/response_datasets/audiomcq.py]: Removed the local conversion — `split_train_validation` now takes `float | int` directly. The coercion itself still has to run (OmegaConf delivers `256.0` from the float-defaulted config node, which `train_test_split` rejects as a fraction), so I centralized it inside `split_train_validation` rather than dropping it. Done in 724d82b.

**Comment by yuekaizhang** [comment]
/ok to test [c16425d](https://github.com/NVIDIA-NeMo/RL/pull/2356/commits/c16425d3aa498a9c37756d991ad156d40c1cb727)

**Comment by yuekaizhang** [comment]
/ok to test [c16425d](https://github.com/NVIDIA-NeMo/RL/pull/2356/commits/c16425d3aa498a9c37756d991ad156d40c1cb727)

**Comment by yuekaizhang** [comment]
/ok to test [c16425d](https://github.com/NVIDIA-NeMo/RL/pull/2356/commits/c16425d3aa498a9c37756d991ad156d40c1cb727)

**Review by yuki-97 — DISMISSED** [review]
lgtm, thanks @yuekaizhang

**Comment by yuekaizhang** [comment]
/ok to test [854d020](https://github.com/NVIDIA-NeMo/RL/pull/2356/commits/854d020727de28cb6376a29ea7605ea8447d368e)

**Comment by yuekaizhang** [comment]
/ok to test [76633ea](https://github.com/NVIDIA-NeMo/RL/pull/2356/commits/76633ea6b890311131d7bc6b2479cd9c4424efa0)

**Comment by yuekaizhang** [comment]
/ok to test [76633ea](https://github.com/NVIDIA-NeMo/RL/pull/2356/commits/76633ea6b890311131d7bc6b2479cd9c4424efa0)

**Review by yuki-97 — APPROVED** [review]
